src/ServiceControl.LearningTransport/LearningTransportNativeQueueLengthProvider.cs

```diff
@@ -0,0 +1,88 @@
+﻿namespace ServiceControl.Transports.LearningTransport
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Linq;
+    using System.Threading.Tasks;
+    using Monitoring;
+    using Monitoring.Infrastructure;
+    using Monitoring.QueueLength;
+
+    public class LearningTransportNativeQueueLengthProvider : INativeQueueLengthProvider
+    {
+        string rootFolder;
+        ConcurrentDictionary<string, LearningTransportQueueLengthTracker> trackers = new ConcurrentDictionary<string, LearningTransportQueueLengthTracker>(StringComparer.InvariantCultureIgnoreCase);
+
+        public Dictionary<EndpointInstanceId, int> SnapshotValues()
+        {
+            return (
+                from t in trackers.Values
```

Comment:
Accessing `Values` or `Keys` in a `ConcurrentDictionary`, locks the whole dictionary. I'd suggest iterating over the whole dictionary `trackers` as enumerator is lock-free.